### PR TITLE
ENT-4353 postgresql migration

### PR DESCRIPTION
**Pull request by Lex-2008**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a pull request! Maybe @craigcomstock can review this?

**Comment by Lex-2008** [comment]
@cf-bottom build, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1967/)

https://ci.cfengine.com/job/pr-pipeline/1967/

**Comment by Lex-2008** [comment]
thankyou!

File: packaging/common/cfengine-hub/postinstall.sh [hunk 1]
```diff
@@ -490,173 +481,227 @@ migrate_db_using_pipe() {
...
+    # 0 - do nothing special (usually pg_upgrade, which is first method, succeeds)
+    # 1 - fail first method (so we get a chance to run second method, migration via pipe)
+    # 2 - fail first two methods (so we get a chance to run third method, migration via dump)
+    # 3 - fail first two methods, and third one on "dumping" stage
+    #     (so we get a chance to see error message with instructions
+    #      how to dump and import manually)
+    # 4 - fail first two methods, and third one on "importing" stage
+    #     (so we get a chance to see error message with instructions
+    #      how to import existing dump file)
+    # 4 - fail first two methods, and third one on "restoring" stage
```
Comment (craigcomstock) [inline packaging/common/cfengine-hub/postinstall.sh]: why two sections with "4 - "?

File: packaging/common/script-templates/script-common.sh [hunk 2]
```diff
@@ -71,3 +71,72 @@ restore_cfengine_state() {
         done
     fi
 }
+
+safe_cp() {
+    # "safe" alternative to `cp`. Tries `cp -al` first, and if it fails - `cp -l`.
```
Comment (craigcomstock) [inline packaging/common/script-templates/script-common.sh]: what if you call safe_cp() ? won't that result in "rm -rf /"!!! ? hilarious right?

File: packaging/common/cfengine-hub/preinstall.sh [hunk 3]
```diff
@@ -137,36 +140,22 @@ if migrating_postgres; then
...
   fi
-  cp -al "$PREFIX/bin" "$BACKUP_DIR"
-  cp -l "$PREFIX/lib"/* "$BACKUP_DIR/lib"
-  cp -al "$PREFIX/lib/postgresql/" "$BACKUP_DIR/lib"
-  cp -al "$PREFIX/share/postgresql/" "$BACKUP_DIR/share"
+  cp -a "$PREFIX/bin" "$BACKUP_DIR"
+  on_files cp "$PREFIX/lib" "$BACKUP_DIR/lib"
+  cp -a "$PREFIX/lib/postgresql/" "$BACKUP_DIR/lib"
+  cp -a "$PREFIX/share/postgresql/" "$BACKUP_DIR/share"
 fi
```
Comment (craigcomstock) [inline packaging/common/cfengine-hub/preinstall.sh]: why not use your safe_cp with all these? trying to do hard-links to save space when it works and failing back to simple cp when on different filesystems?

File: packaging/common/cfengine-hub/postinstall.sh [hunk 4]
```diff
@@ -490,173 +481,227 @@ migrate_db_using_pipe() {
...
-    cp -al "$BACKUP_DIR/bin" "$PREFIX"
-    cp -l "$BACKUP_DIR/lib"/* "$PREFIX/lib"
-    cp -al "$BACKUP_DIR/lib/postgresql/" "$PREFIX/lib"
-    cp -al "$BACKUP_DIR/share/postgresql/" "$PREFIX/share"
+    safe_cp "$BACKUP_DIR" data "$PREFIX/state/pg"
+    cp -a "$BACKUP_DIR/bin" "$PREFIX"
+    on_files cp "$BACKUP_DIR/lib" "$PREFIX/lib"
+    cp -a "$BACKUP_DIR/lib/postgresql/" "$PREFIX/lib"
+    cp -a "$BACKUP_DIR/share/postgresql/" "$PREFIX/share"
     cf_console echo "Dumping old database to SQL file..."
```
Comment (craigcomstock) [inline packaging/common/cfengine-hub/postinstall.sh]: why not use safe_cp for all of these?

File: packaging/common/cfengine-hub/postinstall.sh [hunk 5]
```diff
@@ -490,173 +481,227 @@ migrate_db_using_pipe() {
...
-    rm "$PREFIX/lib"/*
-    mv "$PREFIX/lib/postgresql/" "$BACKUP_DIR.new/lib"
-    mv "$PREFIX/share/postgresql/" "$BACKUP_DIR.new/share"
+    mkdir -p "$PREFIX/lib.new"
+    mkdir -p "$PREFIX/share.new"
+    mv "$PREFIX/bin" "$PREFIX/bin.new"
+    on_files mv "$PREFIX/lib" "$PREFIX/lib.new"
+    mv "$PREFIX/lib/postgresql/" "$PREFIX/lib.new"
+    mv "$PREFIX/share/postgresql/" "$PREFIX/share.new"
     # restore old backup
```
Comment (craigcomstock) [inline packaging/common/cfengine-hub/postinstall.sh]: why not use safe_mv for all of these?

File: packaging/common/cfengine-hub/postinstall.sh (hunk 1 above)
Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh]: You might explain this a bit better. DEBUG variable enables you to debug the script by not executing all of the steps. This allows you to halt the execution at a certain stage and check things out. I wonder though... how do you proceed from there? It's not like the previous steps will re-run cleanly on repetition right?

File: packaging/common/cfengine-hub/postinstall.sh [hunk 6]
```diff
@@ -490,173 +481,227 @@ migrate_db_using_pipe() {
...
+    cf_console echo "Run these commands as 'cfpostgres' user to produce a plaintext dump of the database:"
     cf_console echo
-    cf_console echo "And now installation will proceed with clean (empty) database"
-    (cd /tmp && su cfpostgres -c "$PREFIX/bin/initdb -D $PREFIX/state/pg/data")
+    cf_console echo "LD_LIBRARY_PATH=$BACKUP_DIR/lib/ $BACKUP_DIR/bin/pg_ctl -w -D $BACKUP_DIR/data/ -o '-p 5433' -l /tmp/postgresql-old.log start"
+    cf_console echo "LD_LIBRARY_PATH=$BACKUP_DIR/lib/ $BACKUP_DIR/bin/pg_dumpall --clean --port=5433 >$DUMP_FILENAME"
+    cf_console echo "LD_LIBRARY_PATH=$BACKUP_DIR/lib/ $BACKUP_DIR/bin/pg_ctl -w -D $BACKUP_DIR/data/ -o '-p 5433' stop"
+    cf_console echo
+    cf_console echo "Then, you can import it using these commands (cfengine3 service should be stopped):"
   fi
```
Comment (craigcomstock) [inline packaging/common/cfengine-hub/postinstall.sh]: Why not move the "you can import it..." message outside the if/else and make it the same?

File: packaging/common/cfengine-hub/postinstall.sh [hunk 7]
```diff
@@ -490,173 +481,227 @@ migrate_db_using_pipe() {
...
+  cf_console echo
+  cf_console echo "Alternatively, reinstall CFEngine Enterprise Policy Server version $(cat "$PREFIX/UPGRADED_FROM.txt"),"
+  cf_console echo "and run these commands as 'root' to restore database (cfengine3 service should be stopped):"
+  cf_console echo
+  cf_console echo "rm -rf $PREFIX/state/pg/data"
+  cf_console echo "mv $BACKUP_DIR/data $PREFIX/state/pg/data"
+  cf_console echo
+  cf_console echo "And now installation will proceed with clean (empty) database"
+  (cd /tmp && su cfpostgres -c "$PREFIX/bin/initdb -D $PREFIX/state/pg/data")
+}
```
Comment (craigcomstock) [inline packaging/common/cfengine-hub/postinstall.sh]: I'm a bit worried about the "dead-ness" of all these cf_console echo lines. What if someone changes the code that really does these operations? Could we use the actual code here somehow?

Comment (Lex-2008) [reply packaging/common/cfengine-hub/postinstall.sh]: Hmm... I'll think of a crazy sed oneliner to parse `$0` and output proper lines

Comment (Lex-2008) [reply packaging/common/cfengine-hub/postinstall.sh]: Sorry, I couldn't think of a good enough oneliner. Let's keep it like this and hope ~we'll never need them~ _we'll remember to update them as needed_

**Comment by Lex-2008** [comment]
@cf-bottom build, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1979/)

https://ci.cfengine.com/job/pr-pipeline/1979/

**Comment by Lex-2008** [comment]
@cf-bottom one more build, please!

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1992/)

https://ci.cfengine.com/job/pr-pipeline/1992/

File: packaging/common/cfengine-hub/postinstall.sh [hunk 8]
```diff
@@ -333,6 +333,10 @@ sed -i -e s:INSERT_CERT_HERE:$CFENGINE_MP_CERT:g $PREFIX/httpd/conf/httpd.conf
 sed -i -e s:INSERT_CERT_KEY_HERE:$CFENGINE_MP_KEY:g $PREFIX/httpd/conf/httpd.conf
 sed -i -e s:INSERT_FQDN_HERE:$CFENGINE_LOCALHOST:g $PREFIX/httpd/conf/httpd.conf
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: *finction* in the commit message.

File: packaging/common/cfengine-hub/postinstall.sh [hunk 9]
```diff
@@ -490,173 +481,232 @@ migrate_db_using_pipe() {
...
+  # did pg_dumpall go well?
+  if [ -n "$dump_failed" -o $DEBUG = 3 ]; then
+    cf_console echo "pg_dumpall failed."
+    rm "$BACKUP_DIR/db_dump.sql"
+    return 1
+  fi
+  # did new binaries restored correctly?
+  if [ -n "$restore_failed" -o $DEBUG = 5 ]; then
+    cf_console echo "Failed restoring new binaries."
+    cf_console echo "Your system might be in borked state."
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: *in **a** borked state*

File: packaging/common/cfengine-hub/postinstall.sh [hunk 10]
```diff
@@ -490,173 +481,232 @@ migrate_db_using_pipe() {
...
+
+test -z "$BACKUP_DIR" && BACKUP_DIR=$PREFIX/state/pg/backup
+if [ ! -d $PREFIX/state/pg/data ]; then
+  new_pgconfig_file=`generate_new_postgres_conf`
+  if [ `basename "$new_pgconfig_file"` = "postgresql.conf.cfengine" ]; then
+    pgconfig_type="CFEngine recommended"
+  else
+    pgconfig_type="PostgreSQL default"
   fi
+  init_postgres_dir
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: The `pgconfig` variables are not used here. Please move the code that calls `generate_new_postgres_conf` and sets these variables into `init_postgres_dir`.

File: packaging/common/script-templates/script-common.sh [hunk 11]
```diff
@@ -71,3 +71,75 @@ restore_cfengine_state() {
...
+    # Deletes partially-copied files if copy operation fails.
+    # Args:
+    #   * dir you're copying stuff from
+    #   * name of stuff you're copying (one arg!)
+    #   * dir you're copying stuff to
+    # Exampe: instead of
```
Comment (vpodzime) [inline packaging/common/script-templates/script-common.sh]: *Example*

File: packaging/common/script-templates/script-common.sh [hunk 12]
```diff
@@ -71,3 +71,75 @@ restore_cfengine_state() {
...
+    # perform operation $1 on each file in $2 with optional extra argument $3
+    # Examples:
+    # to copy only files:
+    #   on_files cp "$PREFIX/lib" "$BACKUP_DIR/lib"
+    # to move only files:
+    #   on_files mv "$PREFIX/lib" "$PREFIX/lib.new"
+    # to remove only files:
+    #   on_files rm "$PREFIX/lib"
+    test "$#" -ge 2 || return 2
+    for file in "$2"/*; do
```
Comment (vpodzime) [inline packaging/common/script-templates/script-common.sh]: This won't work on hidden files (starting with a dot).

File: packaging/common/cfengine-hub/postinstall.sh (hunk 10 above)
Comment (Lex-2008) [reply packaging/common/cfengine-hub/postinstall.sh]: But `generate_new_postgres_conf` should be called only _once_, while `init_postgres_dir` can be called _twice_.

**Comment by Lex-2008** [comment]
@cf-bottom build and test

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2017/)

https://ci.cfengine.com/job/pr-pipeline/2017/

File: packaging/common/cfengine-hub/postinstall.sh (hunk 10 above)
Comment (Lex-2008) [reply packaging/common/cfengine-hub/postinstall.sh]: Added a new commit with this change - do you think it looks better this way? If yes, I'll squash it into other one.

Comment (vpodzime) [reply packaging/common/cfengine-hub/postinstall.sh]: Yes, I think this is good.

**Comment by Lex-2008** [comment]
@cf-bottom build, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2034/)

https://ci.cfengine.com/job/pr-pipeline/2034/

**Comment by Lex-2008** [comment]
@cf-bottom build request must you rebase after

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2061/)

https://ci.cfengine.com/job/pr-pipeline/2061/

**Comment by Lex-2008** [comment]
this is WIP because of http://alexey.shpakovsky.ru/en/set--e-inconsistency-inside-blocks.html

**Comment by Lex-2008** [comment]
@cf-bottom build, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2070/)

https://ci.cfengine.com/job/pr-pipeline/2070/

**Comment by Lex-2008** [comment]
@cf-bottom build, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2073/)

https://ci.cfengine.com/job/pr-pipeline/2073/

**Comment by Lex-2008** [comment]
@craigcomstock , @vpodzime : I put amends into "fixups" commit - please review them before I merge 'em into main commit. Also note the difference:
```
set -e
(
false
echo la-la-la
)
```
and
```
set -e
(
false
echo la-la-la
) || uptime
```
in second case, `echo` command gets executed.

**Pull request merged**